contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,193 @@
+
+        uint previousCheckpoint = lastCheckpoint;
+        uint sinceLast = block.timestamp - previousCheckpoint;
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52
+        for (uint i = 0; i < 52; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
+                    /// @dev If no time change since last checkpoint just add new tokens
+                    /// that may have been deposited (same block)
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev In the event that toDistribute contains tokens
+                    /// for multiple weeks we take the remaining portion
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute *
+                            (block.timestamp - previousCheckpoint)) /
+                        sinceLast;
+                }
+                break;
+            } else {
+                /// @dev If passed weeks missed
+                if (sinceLast == 0 && nextWeek == previousCheckpoint) {
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev Store proportion of tokens for this week in the past
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute * (nextWeek - previousCheckpoint)) /
+                        sinceLast;
+                }
+            }
+            previousCheckpoint = nextWeek;
+            thisWeek = nextWeek;
+        }
+        emit CheckpointToken(block.timestamp, toDistribute);
+    }
+
+    /// @notice this function will fetch StakingRewardsV2 to see what their staked balance
+    /// was at the start of the epoch then calculate proportional fees and transfer to user
+    function claimEpoch(address to, uint epochNumber) public override {
+        _checkpointWhenReady();
+        _isEpochReady(epochNumber);
+        if (claimedEpochs[to][epochNumber] == true) {
+            revert CannotClaimTwice();
+        }
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            _startOfEpoch(epochNumber)
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
+        }
+
+        uint256 proportionalFees = calculateEpochFees(to, epochNumber);
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+        claimedEpochs[to][epochNumber] = true;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
```

Comment:
maybe make a constant like `weeksInYear` 
Bad practice having plain numbers floating around

Reply:
I think we told him to do this (over his seconds approach), but yeah I do recommend using constants for anything you **reuse**. For example a constant called `uint YEAR = 52 weeks` if you use it a couple times.

In this case I think you're only using it once so it's up to you if you want to create a constant for it